datafusion/physical-expr/src/aggregate/median.rs

```diff
@@ -0,0 +1,260 @@
+        let mid = len / 2;
+        if len % 2 == 0 {
+            Ok(ScalarValue::$SCALAR_TY(Some(
+                (array.value(mid - 1) + array.value(mid)) / $TWO,
+            )))
+        } else {
+            Ok(ScalarValue::$SCALAR_TY(Some(array.value(mid))))
+        }
+    }};
+}
+
+impl Accumulator for MedianAccumulator {
+    fn state(&self) -> Result<Vec<AggregateState>> {
+        let mut vec: Vec<AggregateState> = self
+            .all_values
+            .iter()
+            .map(|v| AggregateState::Array(v.clone()))
+            .collect();
+        if vec.is_empty() {
+            match self.data_type {
+                DataType::UInt8 => vec.push(AggregateState::Array(Arc::new(
+                    UInt8Array::from_value(0_u8, 0),
+                ))),
+                DataType::UInt16 => vec.push(AggregateState::Array(Arc::new(
+                    UInt16Array::from_value(0_u16, 0),
+                ))),
+                DataType::UInt32 => vec.push(AggregateState::Array(Arc::new(
+                    UInt32Array::from_value(0_u32, 0),
+                ))),
+                DataType::UInt64 => vec.push(AggregateState::Array(Arc::new(
+                    UInt64Array::from_value(0_u64, 0),
+                ))),
+                DataType::Int8 => vec.push(AggregateState::Array(Arc::new(
+                    Int8Array::from_value(0_i8, 0),
+                ))),
+                DataType::Int16 => vec.push(AggregateState::Array(Arc::new(
+                    Int16Array::from_value(0_i16, 0),
+                ))),
+                DataType::Int32 => vec.push(AggregateState::Array(Arc::new(
+                    Int32Array::from_value(0_i32, 0),
+                ))),
+                DataType::Int64 => vec.push(AggregateState::Array(Arc::new(
+                    Int64Array::from_value(0_i64, 0),
+                ))),
+                DataType::Float32 => vec.push(AggregateState::Array(Arc::new(
+                    Float32Array::from_value(0_f32, 0),
+                ))),
+                DataType::Float64 => vec.push(AggregateState::Array(Arc::new(
+                    Float64Array::from_value(0_f64, 0),
+                ))),
+                _ => {
+                    return Err(DataFusionError::Execution(
+                        "unsupported data type for median".to_string(),
+                    ))
+                }
+            }
+        }
+        Ok(vec)
+    }
+
+    fn update_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
+        let x = values[0].clone();
+        self.all_values.extend_from_slice(&[x]);
+        Ok(())
+    }
+
+    fn merge_batch(&mut self, states: &[ArrayRef]) -> Result<()> {
+        for array in states {
+            self.all_values.extend_from_slice(&[array.clone()]);
+        }
+        Ok(())
+    }
+
+    fn evaluate(&self) -> Result<ScalarValue> {
+        match self.all_values[0].data_type() {
+            DataType::Int8 => median!(self, arrow::datatypes::Int8Type, Int8, 2),
```

Comment:
Instead of using a macro here, I wonder if you could use the `concat` and  `take` kernels 

https://docs.rs/arrow/19.0.0/arrow/compute/kernels/concat/index.html
https://docs.rs/arrow/19.0.0/arrow/compute/kernels/take/index.html

Something like (untested):

```rust
let sorted = sort(concat(&self.all_values));
let len = sorted.len();
let mid = len / 2;
if len % 2 == 0 {
  let indexes: UInt64Array = [mid-1, mid].into_iter().collect();
  // 🤔  Not sure how to do an average:
  let values = average(take(sorted, indexes)) 
  ScalarValue::try_from_array(values, 0)
} else {
  ScalarValue::try_from_array(sorted, mid)
} 

```
But the need for an `average` stymies that - though I guess we could implement an `average` kernel in datafusion and then put it back into arrow